### PR TITLE
add support for detecting cycles in the buildgraph

### PR DESCRIPTION
**Pull request by lucasmeijer**
testcoverage for this is on the bee side, I'll open a PR for that shortly

### CODE REVIEW

File: src/LoadOrBuildDag.cpp [hunk 1]
```diff
@@ -117,6 +118,15 @@ bool LoadOrBuildDag(Driver *self, const char *dag_fn)
         return ExitRequestingFrontendRun("%s couldn't be loaded", dag_fn);
     }
 
+    if (DetectCyclicDependencies(self->m_DagData, &self->m_Heap))
```
Comment (bitter) [inline src/LoadOrBuildDag.cpp]: You only need to do this if you get passed a new dag? `self->m_Options.m_DagFileNameJson != nullptr`

File: src/DetectCyclicDependencies.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+                break;
+            cursor = m_ScheduledByWho[cursor];
+        }
+
+        //Ok, now we have the exact path of the cycle, and we can print it out in reverse order:
+        int indent = 0;
+        printf("There is a cycle in the graph produced by your buildprogram:\n");
+        for (int i=exactPathOfCycle.m_Size-1;i>=0;i--)
+        {
+            PrintIndent(indent += 2);
```
Comment (mostynb) [inline src/DetectCyclicDependencies.cpp:124]: maybe just `printf("%*c", indent+=2, ' ');` and delete PrintIndent?

File: src/LoadOrBuildDag.cpp [hunk 3]
```diff
@@ -117,6 +118,15 @@ bool LoadOrBuildDag(Driver *self, const char *dag_fn)
...
+    {
+        MmapFileUnmap(&self->m_DagFile);
+        remove(dag_fn);
+        remove(dagderived_filename);
+        exit(BuildResult::kBuildError);
+        return 0;
```
Comment (mostynb) [inline src/LoadOrBuildDag.cpp]: nitpick: return false

File: src/DetectCyclicDependencies.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+            return true;
+        if (visitDependencies(m_Dag->m_DagNodes[node].m_ToUseDependencies))
+            return true;
+        m_State[node] = VisitState::GuaranteedNoCycles;
+        return false;
+    }
+
+    int FindUnvisitedNode()
+    {
+        for (int i=0; i!=m_NodeCount;i++)
```
Comment (mostynb) [inline src/DetectCyclicDependencies.cpp]: nitpick: add a space before `i++`

**Review by mostynb — APPROVED** [review]

File: src/LoadOrBuildDag.cpp (hunk 1 above)
Comment (lucasmeijer) [reply src/LoadOrBuildDag.cpp]: indeed

File: src/DetectCyclicDependencies.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        m_State[node] = VisitState::BeingVisited;
+        if (visitDependencies(m_Dag->m_DagNodes[node].m_ToBuildDependencies))
+            return true;
+        if (visitDependencies(m_Dag->m_DagNodes[node].m_ToUseDependencies))
+            return true;
+        m_State[node] = VisitState::GuaranteedNoCycles;
+        return false;
+    }
+
+    int FindUnvisitedNode()
```
Comment (richard-fine) [inline src/DetectCyclicDependencies.cpp]: In the worst case - a very flat graph where the majority of nodes will have no dependencies - overall cost of this function will approach O(N^2).

File: src/LoadOrBuildDag.cpp (hunk 1 above)
Comment (lucasmeijer) [reply src/LoadOrBuildDag.cpp]: oh, I see what you mean.  yeah good point.

File: src/DetectCyclicDependencies.cpp (hunk 5 above)
Comment (lucasmeijer) [reply src/DetectCyclicDependencies.cpp]: fixed

**Review by richard-fine — APPROVED** [review]

File: src/LoadOrBuildDag.cpp [hunk 6]
```diff
@@ -117,6 +118,15 @@ bool LoadOrBuildDag(Driver *self, const char *dag_fn)
         return ExitRequestingFrontendRun("%s couldn't be loaded", dag_fn);
     }
 
+    if (DetectCyclicDependencies(self->m_DagData, &self->m_Heap) && self->m_Options.m_DagFileNameJson != nullptr)
```
Comment (bitter) [inline src/LoadOrBuildDag.cpp]: nice try! wrong order tho ;)

**Review by bitter — CHANGES_REQUESTED** [review]

**Review by bitter — APPROVED** [review]

**Pull request merged**